### PR TITLE
Dj rest auth

### LINKED ISSUE #110: [Discussion] New User Registration Confirmation emails & Associated Changes Needed
When a new user signs up send them an email.

Questions:
- Is it just a simple confirmation of signup, does it also need a link to validate the email?
- If a validation link is needed, what's the endpoint for that? Does this feature span frontend and backend repos?
- How is the email delivered: What's your mail transport layer?

### PR DESCRIPTION
**Pull request by BethanyG**
## What type of PR is this? (check all applicable)

- [ ] 🐛 Bug Fix
- [x] ✨ Feature (_**e.g.** API change, enhancement, addition, etc._)
- [x] ♻️ Refactor
- [ ] 📝 Documentation
- [ ] 🔖 Release
- [ ] 🚩 Other (**concerning:**          )

## Context

We wanted to implement a user registration process that incorporated an email and verification of that email.
See issue #110 and discussion #178 for additional context.

Closes #110, #113, #114, #118

## Other Related Tickets & Documents (as needed)

*  [Django-allauth](https://django-allauth.readthedocs.io/en/latest/configuration.html)
*  [dj-rest-auth](https://dj-rest-auth.readthedocs.io/en/latest/installation.html#json-web-token-jwt-support-optional)
*  [SimpleJWT](https://django-rest-framework-simplejwt.readthedocs.io/en/latest/)

## Implementation Details

We had originally planned to use [django-rest-auth](https://github.com/Tivix/django-rest-auth), but it is no longer maintained, and doesn't play nice with `django-allauth` anymore.  `dj-rest-auth` is the currently maintained fork of `django-rest-auth`.  However, `dj-rest-auth` requires the django-simpleJWT library, so the current `django-rest-framework-jwt` library was replaced.

It was also decided that authorization and registration should live under the main API url instead of being "off to the side", so the URLs and code were refactored to support this.  Additionally, settings changes were made to require a validated email upon user registration.  Users will no longer be allowed to login without a "verified" email.

**More refactor/changes may be needed as we refine user profiles and permissions.**

**Obtaining JWT tokens has moved to:**

*   `api/v1/auth/token` (_obtain an access and refresh token pair_), 
*   `api/v1/auth/verify` (_verify the validity of a refresh or access token_), 
*   `api/v1/refresh` (_obtain new access toke by using non-expired refresh token_)

**Registration/Login/Logout are now located at:**

* `api/v1/auth/registration/` (_now triggers a validation email to the email address a user enters._)
* `api/v1/auth/verify-email/` (_to POST a user's HMC email key for validating their email_)
* `api/v1/auth/login/` (_Will require a validated email in order to sign in_)
* `api/v1/auth/logout/` (_Will clear tokens from currently logged in user_)

**User Details and current_user are located at:**

*   `api/v1/auth/user/` (_to view currently logged in User Details_) 
    image
*   `api/v1/auth/current_user` (_to view currently logged in User without email address_)
    image

Since these two are redundant, one may be omitted or changed later in the project.

## New Libraries/Dependancies Introduced (Fill out as needed)

If you have added libraries or other dependancies, please list them (and links to their repos) below:

[djangorestframework-simplejwt](https://github.com/SimpleJWT/django-rest-framework-simplejwt)
[dj-rest-auth==1.1.1](https://github.com/jazzband/dj-rest-auth)
[django-rest-authtoken==2.1.3](https://pypi.org/project/django-rest-authtoken/)

- [x] I've updated **`project/requirements/base.txt`**

## Any new migration files added?

- [x] 👍 yes -- but these should happen in the course of installing the new apps/libraries.
- [ ] 🙅 no, because they aren't needed

## Did you add tests?

**_Code added or changed without test coverage or good reason for exemption won't be approved._**

- [x] 👍 yes - more are needed, but the basics are covered.
- [ ] 🙋 no, because I need help
- [ ] 🙅 no, because they aren't needed

## Did you add documentation?

- [ ] 📜 readme.md
- [ ] 📜 contributing.md
- [ ] 📜 wiki entry
- [X] 🙋 I'd like some additional time to write documentation, and will file a new issue for it
- [ ] 🙅 no documentation needed

### CODE REVIEW

**Comment by lpatmo** [comment]
Yay!!! 👏👏👏 

Excellent work, Bethany! I started reviewing this morning but didn't get a chance to finish testing and looking at everything, but will try to finish up tonight/tomorrow.

One small detail that can be fixed right now: `project/core/templates/account/email/email _confirmation_subject.txt` has an extra space near the underscore.

**Comment by BethanyG** [comment]
Hey @lpatmo - thanks for the sharp-eyed copyedit!  Could you make that suggestion in the PR directly, so we can just accept it here -- or give me a file and line to reference?  Not sure where the URL is is mis-typed?  Thanks!

**Comment by BethanyG** [comment]
OK.  I think I've fixed both the intermittent test failure and the mis-named file.  Test are passing.

File: project/config/settings/base.py [hunk 1]
```diff
@@ -287,21 +290,49 @@
...
+ACCOUNT_EMAIL_REQUIRED = True
+ACCOUNT_EMAIL_VERIFICATION = "mandatory"
+ACCOUNT_AUTHENTICATION_METHOD = "username_email"
+ACCOUNT_CONFIRM_EMAIL_ON_GET = False
+#ACCOUNT_USER_MODEL_USERNAME_FIELD = None
+ACCOUNT_EMAIL_CONFIRMATION_AUTHENTICATED_REDIRECT_URL = None
+#ACCOUNT_EMAIL_CONFIRMATION_ANONYMOUS_REDIRECT_URL = reverse_lazy('account_confirm_complete')
+ACCOUNT_EMAIL_CONFIRMATION_EXPIRE_DAYS = 3
+ACCOUNT_EMAIL_CONFIRMATION_HMAC = True
+ACCOUNT_EMAIL_SUBJECT_PREFIX = "Codebuddies: "
```
Comment (lpatmo) [inline project/config/settings/base.py:313]:
```suggestion
ACCOUNT_EMAIL_SUBJECT_PREFIX = "CodeBuddies: "
```

**Review by lpatmo — APPROVED** [review]
> OK. I think I've fixed both the intermittent test failure and the mis-named file. Test are passing.

Yay, thanks for fixing! 

TIL having `EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'` means the email is "sent" via stdout if I run `docker-compose up` without the `-d` flag :D 

We may need to look into updating the "from" line (a future issue), but I'm very happy with this: 
image

Again, thanks for this big chunk of work! 👏  Feel free to merge in when you're ready. Separately, I filed a FE ticket to look into the changes that need to be made on the FE side: https://github.com/codebuddies/frontend/issues/161

File: project/core/templates/account/email/email _confirmation_subject.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+{% include "account/email/email_confirmation_subject.txt" %}
```
[the remark this answers is not in the document]
Comment (lpatmo) [reply project/core/templates/account/email/email _confirmation_subject.txt]: (Outdated comment from a few days ago when I started reviewing but didn't finish)

File: project/userauth/tests.py [hunk 3]
```diff
@@ -1,180 +1,967 @@
...
+            "username": self.user['username'],
+            "email": self.user['email'],
+            "password1": self.user['password1'],
+            "password2": self.user['password2']
+        }
         response = self.client.post(url, data, format='json')
+
+        #is the email subject what we expect it to be?
+        verify_email_message = mail.outbox[0]
+        self.assertEqual(verify_email_message.subject, 'Codebuddies: Please Confirm Your E-mail Address')
```
Comment (BethanyG) [inline project/userauth/tests.py:98]:
```suggestion
        self.assertEqual(verify_email_message.subject, 'CodeBuddies: Please Confirm Your E-mail Address')
```

Comment (BethanyG) [reply project/userauth/tests.py:98]: Change to email template subject broke test.  Fixing.

**Comment by BethanyG** [comment]
OK.  Mashing the big, green button.  😄

**Pull request merged**